### PR TITLE
PLT-494: PlutusV2 TypedValidators

### PR DESCRIPTION
**Pull request by sjoerdvisscher**
Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [x] Reviewer requested

### CODE REVIEW

**Review by ak3n — APPROVED** [review]
LGTM 👍

File: plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs [hunk 1]
```diff
@@ -752,13 +750,12 @@ resolveScriptTxOut
...
-            { Tx._ciTxOutValidator = (vh, v)
+            { Tx._ciTxOutValidator = (vh, _)
             , Tx._ciTxOutScriptDatum = (dh, d)
             , Tx._ciTxOutValue
             } = do
-    -- first check in the 'ChainIndexTx' for the validator, then
-    -- look for it in the 'slOtherScripts map.
-    (validator, pv) <- maybe (lookupValidator vh) (pure . (, PlutusV1)) v
+    -- Look for the validator in the 'slOtherScripts map so we also get the language.
+    (validator, pv) <- lookupValidator vh
```
Comment (koslambrou) [inline plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs:758]: Would it be benificial to add the PlutusVersion alongside the validator in `ChainIndexTxOut`?

File: plutus-script-utils/src/Plutus/Script/Utils/V1/Typed/Scripts.hs [hunk 2]
```diff
@@ -87,22 +85,6 @@ instance Eq (DatumType a) => Eq (TypedScriptTxOutRef a) where
     tyTxOutRefRef l == tyTxOutRefRef r
       && tyTxOutRefOut l == tyTxOutRefOut r
 
--- | A public-key 'TxOut'. We need this to be sure that it is not a script output.
-newtype PubKeyTxOut = PubKeyTxOut {unPubKeyTxOut :: TxOut}
-  deriving stock (Eq, Generic)
-  -- deriving newtype (FromJSON, ToJSON)
-
--- | Create a 'PubKeyTxOut' from an existing 'TxOut' by checking that it has the right payment type.
-typePubKeyTxOut ::
```
Comment (koslambrou) [inline plutus-script-utils/src/Plutus/Script/Utils/V1/Typed/Scripts.hs:96]: Removed because it's not used anywhere?

**Review by koslambrou — APPROVED** [review]

File: plutus-script-utils/src/Plutus/Script/Utils/V1/Typed/Scripts.hs (hunk 2 above)
Comment (sjoerdvisscher) [reply plutus-script-utils/src/Plutus/Script/Utils/V1/Typed/Scripts.hs:96]: Yes

File: plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs (hunk 1 above)
Comment (sjoerdvisscher) [reply plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs:758]: Just looked at that, it doesn't seem to be possible, at least not easily. We'd also have to store the language in the chain-index.

Comment (koslambrou) [reply plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs:758]: Yeah... quickly looking at cardano-api and ledger, seems like some work and should probably be part of another story. I'm not sure how valuable it is. What do you think?

Comment (sjoerdvisscher) [reply plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs:758]: I won't be surprised that we find out soon that we'll have to do it. I'm fine with waiting until that happens.

**Pull request merged**